### PR TITLE
Bugs fixes for Eigen 3.3, use Eigen CMake target

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This attempts to use Eigen3's CMake target instead of simply specifying the `EIGEN3_INCLUDE_DIR`. 

Previous test failures appear to be related to the lifetime of temporary `Eigen::Vector3d` objects.  

`Eigen::Quaterniond` is a [Fixed-size vectorizable Eigen object](https://eigen.tuxfamily.org/dox/group__TopicFixedSizeVectorizable.html) and so I believe some precautions need to be taken when using them inside [std::vector](https://eigen.tuxfamily.org/dox/group__TopicStlContainers.html#title2)

**To test:**

* Code Review
* Check that the builds still pass

There is no GitHub issue associated with this pull request

**Release Notes** 

*Does not need to be in the release notes.*

Developers should be notified about support for Eigen v3.3.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Beamline/inc/MantidBeamline/ComponentInfo.h [hunk 1]
```diff
@@ -53,7 +53,9 @@ class MANTID_BEAMLINE_DLL ComponentInfo {
       m_componentRanges;
   boost::shared_ptr<const std::vector<size_t>> m_parentIndices;
   Mantid::Kernel::cow_ptr<std::vector<Eigen::Vector3d>> m_positions;
-  Mantid::Kernel::cow_ptr<std::vector<Eigen::Quaterniond>> m_rotations;
+  Mantid::Kernel::cow_ptr<std::vector<
+      Eigen::Quaterniond, Eigen::aligned_allocator<Eigen::Quaterniond>>>
```
Comment (OwenArnold) [inline Framework/Beamline/inc/MantidBeamline/ComponentInfo.h]: After including StdVector -> includes src/StlSupport/StdVector specialisation, I think you will automatically get the right allocator. i.e. no need to express here.

Comment (quantumsteve) [reply Framework/Beamline/inc/MantidBeamline/ComponentInfo.h]: The Eigen documentation says that one needs to **both** the `StdVector` header an the aligned allocator. 

https://eigen.tuxfamily.org/dox/group__TopicStlContainers.html#title2

I tried using the `EIGEN_DEFINE_STL_VECTOR_SPECIALIZATION` that tries to hide the second requirement. Ensuring they are both defined everywhere it is necessary and not defined twice was more trouble than it is worth.

File: Framework/Beamline/CMakeLists.txt [hunk 2]
```diff
@@ -34,12 +34,13 @@ add_library ( Beamline ${SRC_FILES} ${INC_FILES} )
...
                                  INSTALL_RPATH "@loader_path/../MacOS" )
-target_include_directories( Beamline SYSTEM PUBLIC ${EIGEN3_INCLUDE_DIR} )
+
+target_include_directories( Beamline SYSTEM PUBLIC ${Boost_INCLUDE_DIRS})
 
 # Add to the 'Framework' group in VS
 set_property ( TARGET Beamline PROPERTY FOLDER "MantidFramework" )
 
-target_link_libraries ( Beamline LINK_PRIVATE ${TCMALLOC_LIBRARIES_LINKTIME} 
+target_link_libraries ( Beamline LINK_PUBLIC Eigen3::Eigen LINK_PRIVATE ${TCMALLOC_LIBRARIES_LINKTIME}
```
Comment (SimonHeybrock) [inline Framework/Beamline/CMakeLists.txt]: 👍 for more modern use of cmake.

**Review by SimonHeybrock — APPROVED** [review]
Looks good. Alignment can be very important for performance and should maybe be used more within Mantid. 👍

**Pull request merged**